### PR TITLE
[libc++] Add __pointer_int_pair

### PR DESCRIPTION
**Pull request by philnik777**
This is an implementation detail for `move_only_function`.

### CODE REVIEW

**Comment by EricWF** [comment]
I would like to block this until `move_only_function` is up for review.

 I cannot meaningfully review the utility, purpose, and correctness of this class without at least one user.
 
 What is the status of `move_only_function`?

**Comment by ldionne** [comment]
> When in doubt, keep behaviors separate until enough common patterns emerge over time that justify the coupling. On a small scale, managing duplication can be simpler than resolving a premature abstraction’s complexity. In early stages of development, tolerate a little duplication and wait to abstract. [1](https://testing.googleblog.com/2024/05/dont-dry-your-code-prematurely.html)

`pointer_int_pair` is an abstraction that, if we had it, we could have used in a few other places like `optional` and `expected`. Of course, we can't make use of the abstraction until we have an implementation of it.

I generally agree with the statement "don't try to DRY your code too early", however `pointer_int_pair` is a well understood abstraction we're taking from the LLVM code base, we have a few examples of classes that could have used it, and we have an immediately upcoming need for it for `move_only_function`. I don't think this is a case of premature DRYing.

> 
> This is not the only unused abstraction that's been committed to libc++ for move_only_function recently. When looking for the implementation of `move_only_function` I found `__small_buffer` which is similarly unused.

`__small_buffer` was in fact used outside of `move_only_function` after it was introduced, but those uses were reverted due to a "regression" within Google that we still don't have a reproducer for, so we never really picked that up again.

File: libcxx/test/libcxx/utilities/pointer_int_pair/pointer_int_pair.pass.cpp [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+template <class Ptr, class UnderlyingType>
+using single_bit_pair = std::__pointer_int_pair<Ptr, UnderlyingType, std::__integer_width{1}>;
+
+template <class Ptr, class UnderlyingType>
+using two_bit_pair = std::__pointer_int_pair<Ptr, UnderlyingType, std::__integer_width{2}>;
+
+constinit single_bit_pair<int*, size_t> continitiable_pointer_int_pair;
+
+int main(int, char**) {
+  { // __point_int_pair() constructor
```
Comment (ldionne) [inline libcxx/test/libcxx/utilities/pointer_int_pair/pointer_int_pair.pass.cpp:23]:
```suggestion
  { // __pointer_int_pair() constructor
```

File: libcxx/test/libcxx/utilities/pointer_int_pair/static_asserts.verify.cpp [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+TEST_CLANG_DIAGNOSTIC_IGNORED("-Wprivate-header")
+#include <__utility/pointer_int_pair.h>
+TEST_DIAGNOSTIC_POP
+
+// expected-error-re@*:* {{{{static_assert|static assertion}} failed {{.*}}Not enough bits available for requested bit count}}
+std::__pointer_int_pair<char*, size_t, std::__integer_width{2}> ptr1; // expected-note {{here}}
+// expected-error-re@*:* {{{{static_assert|static assertion}} failed {{.*}}_IntType has to be an integral type}}
+std::__pointer_int_pair<int*, int*, std::__integer_width{2}> ptr2; // expected-note {{here}}
+// expected-error-re@*:* {{{{static_assert|static assertion}} failed {{.*}}__pointer_int_pair doesn't work for signed types}}
+std::__pointer_int_pair<int*, signed, std::__integer_width{2}> ptr3; // expected-note {{here}}
```
Comment (ldionne) [inline libcxx/test/libcxx/utilities/pointer_int_pair/static_asserts.verify.cpp:23]:
```suggestion
// expected-error-re@*:* {{Not enough bits available for requested bit count}}
std::__pointer_int_pair<char*, size_t, std::__integer_width{2}> ptr1; // expected-note {{here}}
// expected-error-re@*:* {{_IntType has to be an integral type}}
std::__pointer_int_pair<int*, int*, std::__integer_width{2}> ptr2; // expected-note {{here}}
// expected-error-re@*:* {{__pointer_int_pair doesn't work for signed types}}
std::__pointer_int_pair<int*, signed, std::__integer_width{2}> ptr3; // expected-note {{here}}
```

We could get into the habit of writing those that way to avoid relying too strongly on Clang diagnostic wording.

File: libcxx/include/__utility/pointer_int_pair.h [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+
+_LIBCPP_BEGIN_NAMESPACE_STD
+
+template <class>
+struct _PointerLikeTraits;
+
+template <class _Tp>
+  requires(!is_void_v<_Tp>)
+struct _PointerLikeTraits<_Tp*> {
+  static constexpr size_t __low_bits_available = std::__bit_log2(alignof(_Tp));
```
[the remark this answers is not in the document]
Comment (ldionne) [reply libcxx/include/__utility/pointer_int_pair.h:47]: MTE is an excellent observation. However, it seems they are using the high bits of pointers instead of the low bits like we're doing here. Furthermore, I think it is unlikely that MTE (or anything else) would break us given that LLVM and a bunch of other software is using tricks like these ones. Whatever they do, they have to make it work with code that uses these tricks. So I'm not overly concerned about this.

**Review by ldionne — COMMENTED** [review]
I am generally quite happy with this patch, and IIRC it has been reviewed (and maybe even approved) on Phabricator last year.

Let's see what @EricWF thinks about introducing this now so we can use it immediately after in `move_only_function`, but other than that this LGTM.

**Comment by ldionne** [comment]
@philnik777 Can you upload the patch for `move_only_function` on Github just so we can link to it from here?

**Comment by huixie90** [comment]
Out of interest, how is this useful in the implementation of move_only_function ? I was thinking of

union {
   void* heap_callable_,
   alignas(void*) unsigned char buffer_[N]
   }
   
const vtable* vtable_

Where vtable contains
call
move
Destroy

We don’t seem to need a bool because the vtable function implementation should deal with question whether or not the object is in the buffer or in the heap

**Comment by philnik777** [comment]
`move_only_function` is #94670.

**Comment by EricWF** [comment]
@philnik777 Thanks for putting that up.

I've yet to convince myself that an abstraction this heavy weight is generally useful. I'm worried because it's usefulness is very limited to basically `__pointer_bool_pair`, and in that case we don't need all this bloat.

In other use cases, where the value we pack is more than `0` or `1`, there's a ton of additional concerns and limitations, many of which, like the actual pointer alignment, aren't always computable until runtime when it's too late to change out ABI.

Basically, we need to control own the pointer type & value to prevent issues like with the `[[gnu::packed]]` test that you have. And we need to control the type and value of the integer.

If this is blocking `move_only_function` from moving forward, I'm happy to help unblock that review in particular.

**Comment by EricWF** [comment]
> Out of interest, how is this useful in the implementation of move_only_function ? I was thinking of
> 
> union { void* heap_callable_, alignas(void*) unsigned char buffer_[N] }
> 
> const vtable* vtable_
> 
> Where vtable contains call move Destroy
> 
> We don’t seem to need a bool because the vtable function implementation should deal with question whether or not the object is in the buffer or in the heap

You mean specifically that we can have a null vtable when there's no function contained? That makes sense to me.

**Review by EricWF — CHANGES_REQUESTED** [review]
I think we discovered that this type is not needed (and somewhat harmful to the optimizer)

**Comment by ldionne** [comment]
Closing since inactive: it's not that I think this is a bad idea, I think it's a great idea. But if we revisit this, we should do so with `tagged_pointer` in mind, since that just made it into the standard.

**Comment by philnik777** [comment]
@ldionne I'd really like to keep this open to avoid the code being deleted.

**Comment by ldionne** [comment]
I don't understand? Whether this PR is open or not has no impact on whether the code is deleted unless I missed something? Github still shows the diff, etc.

Generally speaking, we don't keep PRs open if there is no plan to merge them. We've decided we were not going to use `__pointer_int_pair` as a basis for `move_only_function`, so there is no planned usage of this code at the moment.

Don't get me wrong, I think this code was really neat, but I just don't think that's a sufficient reason to keep the PR open.

**Comment by philnik777** [comment]
> I don't understand? Whether this PR is open or not has no impact on whether the code is deleted unless I missed something? Github still shows the diff, etc.

This branch is in the monorepo and branches in there get deleted. I don't know whether that's the case for diffs eventually as well, but I also don't really want to find out.

> Generally speaking, we don't keep PRs open if there is no plan to merge them. We've decided we were not going to use `__pointer_int_pair` as a basis for `move_only_function`, so there is no planned usage of this code at the moment.
> 
> Don't get me wrong, I think this code was really neat, but I just don't think that's a sufficient reason to keep the PR open.

I do plan to eventually use it in `__tree`. I simply didn't get to it yet.

**Comment by ldionne** [comment]
I'll make this a draft then so it at least doesn't clutter the list of PRs.
